File: 304_Range_Sum_Query_2D_Immutable/sumRegion.py
class NumMatrix(object):
    def __init__(self, matrix):
        """
        Initialize your data structure here.
        :type matrix: List[List[int]]
        """
        if not matrix:
            self.sumMatrix = []
        else:
            r_length = len(matrix[0]) + 1
            c_length = len(matrix)
            self.sumMatrix = [[0] * r_length for _ in xrange(c_length)]
            for i in xrange(c_length):
                for j in xrange(1, r_length):
                    self.sumMatrix[i][j] = self.sumMatrix[i][j - 1] + matrix[i][j - 1]

    def sumRegion(self, row1, col1, row2, col2):
        """
        sum of elements matrix[(row1, col1)..(row2, col2)], inclusive
        :type row1: int
        :type col1: int
        :type row2: int
        :type col2: int
        :rtype: int
        """
        if not self.sumMatrix: return 0
        res = 0
        for i in xrange(row1, row2 + 1):
            res += self.sumMatrix[i][col2 + 1] - self.sumMatrix[i][col1]
        return res
"""
matrix = [[3, 0, 1, 4, 2],
          [5, 6, 3, 2, 1],
          [1, 2, 0, 1, 5],
          [4, 1, 0, 1, 7],
          [1, 0, 3, 0, 5]]
"""
matrix = []
numMatrix = NumMatrix(matrix)
print numMatrix.sumRegion(2, 1, 4, 3) 

File: 304_Range_Sum_Query_2D_Immutable/sumRegion.cpp
class NumMatrix {
public:
    NumMatrix(vector<vector<int>> &matrix) {
        // coner case: matrix = NULL
        if(matrix.size() != 0) {
            int n = matrix.size();
            int m = matrix[0].size();
            for(int i = 0; i < matrix.size(); i++) {
                vector<int> sum(m + 1, 0);
                for(int j = 1; j <= m; j++) {
                    sum[j] = sum[j - 1] + matrix[i][j - 1];
                }
                sums.push_back(sum);
            }
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2) {
        if(sums.size() == 0) { return 0; }
        int res = 0;

        for(int i = row1; i <= row2; i++) {
            res += (sums[i][col2 + 1] - sums[i][col1]);
        }

        return res;
    }

private:
    vector<vector<int> > sums;

};